File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiStatusBar.cpp
#include "StdAfx.h"
#include "UiStatusBar.h"

#define ThisMethod($x) &UiStatusBar::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiStatusBar::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiStatusBar> c_obj;
	}

	void UiStatusBar::DefineControl()
	{
		AutoAddMethod( SetPart );
		AutoAddMethod( GetPart, WrapObjectGeneric );
		AutoAddMethod( SetWidth );
		AutoAddMethod( GetWidth, WrapObjectGeneric );
		AutoAddMethod( GetWidthReal, WrapObjectGeneric );
		AutoAddMethod( SetMaxWidth );
		AutoAddMethod( GetMaxWidth, WrapObjectGeneric );
		AutoAddMethod( SetReducePriority );
		AutoAddMethod( GetReducePriority, WrapObjectGeneric );
		AutoAddMethod( SetText );
		AutoAddMethod( GetText, WrapObjectGeneric );
		AutoAddMethod( SetPartToolTip );
		AutoAddMethod( SetAlign );
		AutoAddMethod( GetAlign, WrapObjectGeneric );
		AutoAddMethod( SetIcon );
		AutoAddMethod( GetIcon, WrapObjectGeneric );
		AutoAddMethod( SetClickable );
		AutoAddMethod( GetClickable, WrapObjectGeneric );
		AutoAddMethod( SetContent );
		AutoAddMethod( GetContent, WrapObjectGeneric );
		AutoAddMethod( SetContentMargin );
		AutoAddMethod( GetContentMargin, WrapObjectGeneric );
		AutoAddMethod( HitTest, WrapObjectGeneric );
		AutoAddMethod( GetItemRect, WrapObjectGeneric );
		AutoAddMethod( SetBackground );
		AutoAddMethod( GetBackground, WrapObjectGeneric );
		AutoAddMethod( SetBackColor );
		AutoAddMethod( GetBackColor, WrapObjectGeneric );
		AutoAddMethod( OnClick, WrapObjectGeneric );
	}

	U1 UiStatusBar::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		GetControlTyped().SetIconManager( p->PublicCloneIconManager() );

		return true;
	}

	void UiStatusBar::__OnClick( Ui::IStatusBar & sender, S32 nIndex )
	{
		m_OnClick( this, nIndex );
	}

	UiStatusBar * UiStatusBar::SetPart( const List<WrapData<Ui::DpiSize>>& vPartWidth )
	{
		List<Ui::DpiSize> v( vPartWidth.Size() );
		for ( auto& i : vPartWidth.RangeIndexAll() )
			v[i.Index] = *i;
		GetControlTyped().SetPart( v.Data(), (U32) v.Size() );
		return this;
	}

	List<WrapData<Ui::DpiSize>> UiStatusBar::GetPart() const
	{
		List<Ui::DpiSize> v( GetControlTyped().GetPart( nullptr, 0 ) );
		if ( !v.IsEmpty() )
		{
			GetControlTyped().GetPart( v.Data(), (U32) v.Size() );
			List<WrapData<Ui::DpiSize>> vr( v.Size() );
			for ( auto& i : v.RangeIndexAll() )
				vr[i.Index] = *i;
			return std::move( vr );
		}
		return {};
	}

	UiStatusBar * UiStatusBar::OnClick( OnClick_t && fn ) { m_OnClick = SetEventCallback<Ui::IStatusBar::OnClick>( std::move( fn ), MakeThisFunc( __OnClick ) ); return this; }

}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiStatusBar.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { Window } from "./UiWindow";
import { DpiSize, AlignType, IconCache, Rect, StringKey } from "../UiCommon";
import { Vec2, Vec4 } from "../../Math/Vector";

export interface IStatusBar extends IControl {
    new (window: Window, key?: StringKey): IStatusBar;

    SetPart(vPartWidth: DpiSize[]): StatusBar;
    GetPart(): DpiSize[];
    SetWidth(nIndex: number, nWidth: DpiSize): StatusBar;
    GetWidth(nIndex: number): DpiSize;
    GetWidthReal(nIndex: number): number;
    SetMaxWidth(nIndex: number, nWidth: DpiSize): StatusBar;
    GetMaxWidth(nIndex: number): DpiSize;
    SetReducePriority(nIndex: number, nPriority: number): StatusBar;
    GetReducePriority(nIndex: number): number;
    SetText(nIndex: number, s: string): StatusBar;
    GetText(nIndex: number): string;
    SetPartToolTip(nIndex: number, s: string): StatusBar;
    SetAlign(nIndex: number, nType: AlignType): StatusBar;
    GetAlign(nIndex: number): AlignType;
    SetIcon(nIndex: number, nIcon: IconCache): StatusBar;
    GetIcon(nIndex: number): IconCache;
    SetClickable(nIndex: number, bClickable: boolean): StatusBar;
    GetClickable(nIndex: number): boolean;
    SetContent(nIndex: number, pControl: IControl): IControl;
    GetContent(nIndex: number): IControl;
    SetContentMargin(nIndex: number, nMargin: number): StatusBar;
    GetContentMargin(nIndex: number): number;
    HitTest(pt: Vec2, bIncludeUnclickable: boolean): number;
    GetItemRect(nIndex: number): Rect;

    SetBackground(b: boolean): StatusBar;
    GetBackground(): boolean;

    SetBackColor(vColor: Vec4): StatusBar;
    GetBackColor(): Vec4;

    OnClick(fn: (sender: StatusBar, nIndex: number) => void): StatusBar;
}

export class StatusBar extends (AveLib.UiStatusBar as IStatusBar) {
    private m_Content: Array<IControl>;
    SetPart(vPartWidth: DpiSize[]) {
        this.m_Content = null;
        if (vPartWidth.length > 0)
            this.m_Content = new Array(vPartWidth.length);
        return super.SetPart(vPartWidth);
    }

    SetContent(nIndex: number, pControl: IControl) {
        if (this.m_Content && nIndex < this.m_Content.length)
            this.m_Content[nIndex] = pControl;
        return super.SetContent(nIndex, pControl);
    }

    GetItemRect(nIndex: number): Rect {
        return Rect.FromNative(super.GetItemRect(nIndex));
    }
}
